File: modules/driver-fatigue-vision/android/src/main/java/expo/modules/driverfatiguevision/DriverFatigueVisionModule.kt
package expo.modules.driverfatiguevision

import android.Manifest
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class DriverFatigueVisionModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("DriverFatigueVision")

    Function("getCameraPermissionStatus") {
      cameraPermissionStatus()
    }

    AsyncFunction("requestCameraPermission") {
      cameraPermissionStatus()
    }

    View(DriverFatigueVisionView::class) {
      Events("onFrameResult", "onError", "onStatusChange")

      Prop("active") { view: DriverFatigueVisionView, active: Boolean ->
        view.setActive(active)
      }

      Prop("targetFps") { view: DriverFatigueVisionView, targetFps: Int ->
        view.setTargetFps(targetFps)
      }
    }
  }

  private fun cameraPermissionStatus(): String {
    val context = appContext.reactContext ?: return "undetermined"
    return if (
      ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA) ==
        PackageManager.PERMISSION_GRANTED
    ) {
      "granted"
    } else {
      "undetermined"
    }
  }
}

File: modules/driver-fatigue-vision/ios/DriverFatigueVisionModule.swift
import ExpoModulesCore
import AVFoundation

public class DriverFatigueVisionModule: Module {
  public func definition() -> ModuleDefinition {
    Name("DriverFatigueVision")

    Function("getCameraPermissionStatus") {
      Self.cameraPermissionStatus()
    }

    AsyncFunction("requestCameraPermission") { (promise: Promise) in
      let status = AVCaptureDevice.authorizationStatus(for: .video)
      guard status == .notDetermined else {
        promise.resolve(Self.cameraPermissionStatus())
        return
      }

      AVCaptureDevice.requestAccess(for: .video) { granted in
        promise.resolve(granted ? "granted" : "denied")
      }
    }

    View(DriverFatigueVisionView.self) {
      Events("onFrameResult", "onError", "onStatusChange")

      Prop("active") { (view, active: Bool) in
        view.setActive(active)
      }

      Prop("targetFps") { (view, targetFps: Int) in
        view.setTargetFps(targetFps)
      }
    }
  }

  private static func cameraPermissionStatus() -> String {
    switch AVCaptureDevice.authorizationStatus(for: .video) {
    case .authorized:
      return "granted"
    case .denied:
      return "denied"
    case .restricted:
      return "restricted"
    case .notDetermined:
      return "undetermined"
    @unknown default:
      return "undetermined"
    }
  }
}
